Archipelago.MultiClient.Net/Cache/DataPackageFileSystemCache.cs

```diff
@@ -0,0 +1,171 @@
+using Archipelago.MultiClient.Net.Packets;
+using Newtonsoft.Json;
+
+namespace Archipelago.MultiClient.Net.Cache
+{
+    internal class DataPackageFileSystemCache : IDataPackageCache
+    {
+        private const string DataPackageFileName = "datapackagecache.archipelagocache";
+        private readonly ArchipelagoSocketHelper socket;
+        private string CacheFolder = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+        private RoomInfoPacket roomInfoPacket;
+        private DataPackage dataPackage;
+
+        private object fileAccessLockObject = new object();
+
+        public DataPackageFileSystemCache(ArchipelagoSocketHelper socket)
+        {
+            this.socket = socket;
+
+            socket.PacketReceived += Socket_PacketReceived;
+        }
+
+        private void Socket_PacketReceived(ArchipelagoPacketBase packet)
+        {
+            if (packet.PacketType == ArchipelagoPacketType.RoomInfo)
+            {
+                roomInfoPacket = (RoomInfoPacket)packet;
+                var invalidated = GetCacheInvalidatedGames(roomInfoPacket);
+                
+                if (invalidated.Any())
+                {
+                    var exclusions = roomInfoPacket.DataPackageVersions.Select(x => x.Key).Except(invalidated);
+
+                    socket.SendPacket(new GetDataPackagePacket()
+                    {
+                        Exclusions = exclusions.ToList()
+                    });
+                }
+            }
+            else if (packet.PacketType == ArchipelagoPacketType.DataPackage)
+            {
+                var packagePacket = (DataPackagePacket)packet;
+                SaveDataPackageToCache(packagePacket.DataPackage);
+            }
+        }
+
+        public bool TryGetDataPackageFromCache(out DataPackage package)
+        {
+            if (dataPackage != null)
+            {
+                package = dataPackage;
+                return true;
+            }
+            lock (fileAccessLockObject)
+            { 
+                var dataPackagePath = Path.Combine(CacheFolder, DataPackageFileName);
+                if (File.Exists(dataPackagePath))
+                {
+                    try
+                    {
+                        string fileText = File.ReadAllText(dataPackagePath);
+                        package = JsonConvert.DeserializeObject<DataPackage>(fileText);
+                        dataPackage = package;
+                        return true;
+                    }
+                    catch (Exception e)
+                    {
+                        throw new CacheLoadFailureException("Could not load data package cache from file system.", e);
+                    }
+                }
+                else
+                {
+                    package = null;
+                    return false;
+                }
+            }
+        }
+
+        public bool SaveDataPackageToCache(DataPackage package)
+        {
```

Comment:
Not sure if we should handle this on Saving or deciding which games are to invalidated when we receive the roominfo. but we need to handle the case where a game version is 0, right now there is a change your local cached version is 0 and it wont update until the version is actually changed

Reply:
Will handle datapackage version 0 somewhere. Thanks!